agent/handlers/v2/response.go

```diff
@@ -144,32 +144,41 @@ func NewTaskResponse(
 	}
 
 	if propagateTags {
-		propagateTagsToMetadata(state, ecsClient, containerInstanceArn, taskARN, resp)
+		if err := propagateTagsToMetadata(ecsClient, containerInstanceArn, task.Arn, resp); err != nil {
+			return nil, err
+		}
 	}
 
 	return resp, nil
 }
 
-func propagateTagsToMetadata(state dockerstate.TaskEngineState, ecsClient api.ECSClient, containerInstanceArn, taskARN string, resp *TaskResponse) {
+func propagateTagsToMetadata(
+	ecsClient api.ECSClient,
+	containerInstanceArn,
+	taskARN string,
+	resp *TaskResponse,
+) error {
 	containerInstanceTags, err := ecsClient.GetResourceTags(containerInstanceArn)
-	if err == nil {
-		resp.ContainerInstanceTags = make(map[string]string)
-		for _, tag := range containerInstanceTags {
-			resp.ContainerInstanceTags[*tag.Key] = *tag.Value
-		}
-	} else {
-		seelog.Errorf("Could not get container instance tags for %s: %s", containerInstanceArn, err.Error())
+	if err != nil {
```

Comment:
Just noticed that this is also going to affect the behavior of `v2` endpoint, correct?
if so, is that expected?

Reply:
Yes, this changes the behavior of v2 endpoint. As no change has been made on `V2 TaskContainerMetadataHandler`, if an error is returned from the ECS API, the response of TMDE v2 path`/v2/metadataWithTags` will remain the same, as the following code snippets from function [WriteTaskMetadataResponse](https://github.com/aws/amazon-ecs-agent/blob/master/agent/handlers/v2/task_container_metadata_handler.go#L92) in task_container_metadata_handler.go
```
taskResponse, err := NewTaskResponse(taskARN, state, ecsClient, cluster, az, containerInstanceArn, propagateTags, false)
if err != nil {
	errResponseJSON, err := json.Marshal("Unable to generate metadata for task: '" + taskARN + "'")
	if e := utils.WriteResponseIfMarshalError(w, err); e != nil {
		return
	}
	utils.WriteJSONToResponse(w, http.StatusBadRequest, errResponseJSON, utils.RequestTypeTaskMetadata)
	return
}
```

Reply:
The difference is that, in your snipped above, `NewTaskResponse` did not return error if the tags failed, but now it will. It would be worth to discuss this a bit more offline with people that have more context on v2 to see if this is a concern nor not.

Reply:
Should we use a new flag, like includeECSError, for v3 and v4, and only return error info for them?

In amazon-ecs-agent/agent/handlers/v2/response.go
```
// NewTaskResponse creates a new response object for the task
func NewTaskResponse(
	taskARN string,
	state dockerstate.TaskEngineState,
	ecsClient api.ECSClient,
	cluster string,
	az string,
	containerInstanceArn string,
	propagateTags bool,
	includeV4Metadata bool,
        includeECSError bool,
) (*TaskResponse, error) {...}
```

```
func propagateTagsToMetadata(
	ecsClient api.ECSClient,
	containerInstanceArn,
	taskARN string,
	resp *TaskResponse,
        includeECSError bool,
) error {
	containerInstanceTags, err := ecsClient.GetResourceTags(containerInstanceArn)
	if err == nil {
		...
	} else {
		if !includeECSError {
			seelog.Errorf("Could not get container instance tags for %s: %s", containerInstanceArn, err.Error())
		} else {
			return err
		}
	}

	taskTags, err := ecsClient.GetResourceTags(taskARN)
	if err == nil {
		...
	} else {
		if !includeECSError {
			seelog.Errorf("Could not get task tags for %s: %s", taskARN, err.Error())
		} else {
			return err
		}
	}
        return nil
}
```

Reply:
I would either vote for changing this behavior for v2, v3, v4 together(since this is an undocumented endpoint), or just change v4 alone with a flag to throw an error(if v4) since we are updating just the latest going forward :)